packages/api/src/controllers/access-control.ts

```
@@ -50,73 +57,184 @@ app.post(
+          throw new NotFoundError("Content not found");
+        }
+
+        if (signingKey.disabled || signingKey.deleted) {
+          const signingKeyLog = JSON.stringify(JSON.stringify(signingKey));
+          console.log(`
+            access-control: gate: disallowing playback for contentId=${content.id} with playbackId=${playbackId} the public key pub=${signingKey.id} is disabled or deleted, signingKey=${signingKeyLog}
+          `);
+          throw new ForbiddenError("The public key is disabled or deleted");
+        }
+
+        tracking.recordSigningKeyValidation(signingKey.id);
+        res.set("Cache-Control", "max-age=120,stale-while-revalidate=600");
+        res.status(204);
+        return res.end();
+      case "webhook":
+        if (!req.body.accessKey) {
+          console.log(`
+            access-control: gate: no accessKey provided for playbackId ${playbackId}, disallowing playback
+          `);
+          throw new ForbiddenError(
+            "Content is gated and requires an access key"
+          );
+        }
+        const webhook = await db.webhook.get(content.playbackPolicy.webhookId);
+        if (!webhook) {
+          console.log(`
+            access-control: gate: content with playbackId ${playbackId} is gated but corresponding webhook not found for webhookId ${content.playbackPolicy.webhookId}, disallowing playback
+          `);
+          throw new ForbiddenError(
+            "Content is gated and corresponding webhook not found"
+          );
+        }
+        const statusCode = await fireGateWebhook(
+          webhook,
+          content.playbackPolicy,
+          req.body.accessKey
         );
-      }
-
-      if (signingKeyOutput.length > 1) {
-        let collisionKeys = JSON.stringify(signingKeyOutput);
+        switch (statusCode) {
+          case 200:
+            res.set("Cache-Control", "max-age=120,stale-while-revalidate=600");
+            res.status(204);
+            return res.end();
+          case 0:
```

How can a 0 status code happen? 0o

---

I am specifically setting the status code at 0 when I am not able to receive it (e.g. timeouts), as it's usually done by xhrxmlhttprequest, but I can change it and make it undefined instead in that case!

---

https://fetch.spec.whatwg.org/#concept-network-error

---

Hm, not a fan of returning a dummy value on an exceptional case, that's what exceptions are for and both provide more info (error msg) + avoid the need of understanding how such dummy values are created (which requires reading the sub-function code or docs, which we don't usually write). 

But that's too low-level and doesn't matter much anyway. If you're more comfortable following that pattern, just keep it!

---

I am keeping the 0 status code in the ts logic and returning 502 Bad Gate(way) to Catalyst, FYI

---

Sounds good!